### ISSUE TITLE
Set LCD geometry for subpixel rendering

### ISSUE THREAD

**Issue by Qrox** [libsdl-org/SDL_ttf#296]
Currently, SDL_ttf uses the Harmony algorithm with default LCD geometry (uniform RGB subpixels in the horizontal direction) when doing subpixel rendering. According to [freetype docs](https://freetype.org/freetype2/docs/reference/ft2-lcd_rendering.html),

> Harmony with default LCD geometry is equivalent to ClearType with light filter.

which apparently does not work for screens with a different subpixel layout, as evidenced by [this post](https://www.reddit.com/r/Monitors/comments/w9qm0y/actual_fix_for_the_aw3423dw_subpixel_layouttext/).

Freetype has a function [`FT_Library_SetLcdGeometry`](https://freetype.org/freetype2/docs/reference/ft2-lcd_rendering.html#ft_library_setlcdgeometry) for setting the geometry but it requires passing in the library instance which is a static variable in SDL_ttf, so it is currently impossible to set the LCD geometry when using SDL_ttf.

It would be nice if SDL_ttf can expose this function to achieve the best rendering quality on non-conventional screens.

**Comment by 1bsyl** [comment]
Yes, I think this make sense to expose a wrapper to be able to use this function !  
feel free to make a PR for it

**Comment by 1bsyl** [comment]
I think adding / exposing the freetype function is the way to go:
 TTF_GetSubpixelMode()
 TTF_SetLcdFilter()
 TTF_SetLcdFilterWeights()
 TTF_SetLcdGeometry()

@Qrox 
hey, can you try this patch, to customize the subpixel rendering ?

**Comment by Qrox** [comment]
@1bsyl Thanks for the update. The patch looks good to me but unfortunately I do not have time to test it right now. I can provide some observation that might make sense to note in the documentation though.

1. `FT_Library_SetLcdGeometry`'s y coordinates are in reverse from the usual SDL notation (larger = upper as opposed to larger = lower).
2. The current code does not clear the glyph cache of created SDL_ttf fonts when changing the filter/geometry, but clearing the glyph cache would require keeping a list of created fonts since the filter/geometry are global settings.
3. The current implementation of `TTF_GetSubpixelMode` may reset the geometry and/or filter.

**Comment by 1bsyl** [comment]
update: 
- SDLCALL
- const int param
- rename param to pixel_coordinates
- add comment for FT_Library_SetLcdGeometry: 
  since, this is just a stub to FT library ... just add a remark in the header...

- I think TTF_GetSubpixelMode() is ok. because "done" and "mode" are static var. so the detection is done once. then the cached value is returned ?

**Comment by 1bsyl** [comment]
indeed: 
The current code does not clear the glyph cache of created SDL_ttf fonts when changing the filter/geometry, but clearing the glyph cache would require keeping a list of created fonts since the filter/geometry are global settings.

**Comment by 1bsyl** [comment]
should we maintain an internal list of open fonts or just document the behavior :/

**Comment by Qrox** [comment]
There's also the scenario of multiple monitors with different subpixel layouts, and not clearing the glyph cache can enable the user to create different versions of a font to apply to different monitors. But it sounds a bit hacky and I suppose the proper way is to ask freetype to add a per-font layout setting.

**Comment by 1bsyl** [comment]
maybe, we should tweak those new TTF_SetLcd function to take a TTF_Font in parameter, 
We only apply the filter setting the current font (and set back to default for other fonts), and we handle the lcd filter switching internally so that each font is rendered with its own filter.

**Comment by 1bsyl** [comment]
I've update the PR so that the setting are per-font, 
https://github.com/libsdl-org/SDL_ttf/pull/302

re-apply the setting if they have changed, flush the cache.
ok, compiles, but not tested at all :)

**Comment by 1bsyl** [comment]
@Qrox 
it would be nice if you could give a try when you have some time
thanks

**Comment by Qrox** [comment]
Sure, I have some in progress changes to `showfont.c` to test LCD rendering mode, when I have more time I'll test your PR with those changes and create a PR.
